### PR TITLE
Mark known test failures

### PR DESCRIPTION
**Pull request by macro1**
Marks known test failures tied to #2131 as we do not have a strategy for remediating.

##### Contributor checklist

- [ ] Included tests for the changes.
- [ ] PR title is short, clear, and ready to be included in the user-facing changelog.

##### Maintainer checklist

- [ ] Verified one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by webknjaz** [comment]
IIRC, the fix is #2106. I'd rather accept it instead.

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:3458]: I don't think it's acceptable to both mark the test as xfail and make changes to it simultaneously.

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:1482]: The message should be clear about what versions are bad so that it's more obvious when it can be dropped. This is a part of the test run output to the terminal, saying that some arbitrary pip version that was newer at an arbitrary point in time in the past is unhelpful.

File: .pre-commit-config.yaml
Comment (webknjaz) [inline .pre-commit-config.yaml:38]: I don't think this should be a part of this PR. If we want to merge a few PRs simultaneously, it's best to make a cumulative PR with a real/natural merge commit combining several branches into `main` so that merging would pick up all the PRs individually.

**Review by webknjaz — CHANGES_REQUESTED** [review]
While I don't think that the tests should be suppressed at all, if we go this way, the places mentioned below must be fixed first.

File: tests/test_cli_compile.py
Comment (macro1) [reply tests/test_cli_compile.py:3458]: the test above marked with xfail is this one:
`test_preserve_via_requirements_constrained_dependencies_when_run_twice()`
whereas the test being adjusted is this one:
`test_compile_recursive_extras_build_targets()`

i can see how it may look like the same test though

File: .pre-commit-config.yaml
Comment (macro1) [reply .pre-commit-config.yaml:38]: that's fine. my goal here was to get CI passing again, which included this change. it is currently brought in with a merge, but i'm happy to rebase once the other pr is in.

**Comment by macro1** [comment]
@webknjaz we either suppress or halt development on the project. there's no proposed solution for #2110

File: tests/test_cli_compile.py
Comment (webknjaz) [reply tests/test_cli_compile.py:3458]: Oh, that's what I missed. In that case, this change is incorrect, as it masks the problem that #2106 is attempting to fix.

**Review by webknjaz — CHANGES_REQUESTED** [review]
> we either suppress or halt development on the project. there's no proposed solution for #2131

I guess you could work on the solution to that issue with a few temporary commits that unblock the CI for you and could be dropped later.

Meanwhile, the correct solution to the failing tests is #2106. Fixing it is what has to happen first.

I'm blocking this because the change at https://github.com/jazzband/pip-tools/pull/2137/files#r1867006850 is harmful in that it contributes to shoving the problem under the rug, which is a big deal.

Since nobody seems to be attempting to look into completing #2106, I'll try to check it myself later today.

**Comment by macro1** [comment]
alright. so we'll rely on #2106 for the wheel expected in output issue, #2134 for the pre-commit issue (the mypy step no longer supports 3.8 runtime). in this PR, i'll narrow it down to the known issue introduced in pip 24.3

i'm curious about #2106 and the problems we would potentially be shoving beneath the rug, so i may comment there later when i have more time

**Comment by webknjaz** [comment]
I've converted this to draft to prevent accidental merging ahead of time.

**Comment by macro1** [comment]
@webknjaz currently this PR is blocked by your request for changes, and now it is also a draft, so i don't think it's going anywhere

what needs to be merged ahead of this PR? the changes have been reduced to marking specific tests as expected failures when using a pip version of 24.3 or higher, which will remain true till there's a solution for #2131

**Comment by webknjaz** [comment]
@macro1 in #2106, I'm experimenting with making the testing more predictable by pinning pip more aggressively in the test infra (https://github.com/jazzband/pip-tools/pull/2106#issuecomment-2516076142). I recall that PR failing weirdly on Windows, and now that's not exactly the case.

Now that I identified this, I'm going to see if I can change that setup together with #2134 and make pip updates in CI more controlled.

**Comment by webknjaz** [comment]
@macro1 this is not technically necessary anymore. I've still clicked “rebase” to get the CI statuses to me green. Though, perhaps this should be updated together with bumping the pip version that I pinned in another PR. And with the corresponding fixes instead of disabling the tests...

**Comment by webknjaz** [comment]
@macro1 let me know if you think this should be merged anyway.

**Comment by macro1** [comment]
@webknjaz i think i mentioned before there's no current solution for the failing tests this PR is set to mark as known failures. you're currently skipping the tests by limiting the versions of pip we run any tests for, whereas this PR was specifically targeting the known failures.

imo introducing the idea of supported versions of pip runs contrary to how people expect to use this project, but it's a difference of philosophy and probably won't matter in the long-run.

i disagree and do not think it's my choice on whether to merge these changes in at this point. we now have the concept of supported pip versions.. if that is sufficient, that is the approach you have led us to. in my opinion, you've made the choice for us on which approach to take

i do want you to know i'm really happy CI is able to produce green statuses again, and it's awesome there are active contributors here

**Comment by webknjaz** [comment]
@macro1 I asked you if you wanted to merge because I think your opinion is important too. I'm kinda open to it but it might make sense to couple it with bumping pip.

As for "supported", Hugo raised a concern in another PR. We talked about renaming it but postponed, prioritizing making the PR green. It turned out to be confusing but I did not mean to imply that everything else is unsupported. I just wanted to have something that the PR authors would be able to rely on and could be bumped in a controlled and transparent way.
In the metadata, this is not limited.
We should probably rename it to "latest tested".

**Comment by macro1** [comment]
including this change when adding the other pip versions to testing makes sense to me

**Pull request closed without merging**
